### PR TITLE
Fixes #16256: Deadlock on compliance on Rudder 5.0

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/16256

### CODE REVIEW

**Comment by ncharles** [comment]
do not merge for now

**Comment by ncharles** [comment]
let's wait for one night to see if lock appear again

**Comment by ncharles** [comment]
This works fine, even if it makes things a bit slower

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala [hunk 1]
```diff
@@ -406,43 +409,51 @@ trait DefaultFindRuleNodeStatusReports extends ReportingService {
...
+        _ = TimingDebugLogger.trace(s"Compliance: get Execution Reports in batch for ${runInfos.size} runInfos: ${(t1 - t0)/1000}µs")
+      } yield {
+        val t2 = System.nanoTime()
+        //we want to have nodeStatus for all asked node, not only the ones with reports
+        val nodeStatusReports = runBatch.map { case (nodeId, runInfo) =>
+          val status = ExecutionBatch.getNodeStatusReports(nodeId, runInfo, reports.getOrElse(nodeId, Seq()), unexpectedInterpretation)
+          (status.nodeId, status)
+        }
+        val t3 = System.nanoTime()
+        u2 += (t2 - t1)
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala]: shouldn't it be `t3-t2` here ?

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala [hunk 2]
```diff
@@ -406,43 +409,51 @@ trait DefaultFindRuleNodeStatusReports extends ReportingService {
...
-                                    }
-      t1      =  System.currentTimeMillis
-      _       =  TimingDebugLogger.trace(s"Compliance: get Execution Reports for ${runInfos.size} runInfos: ${t1-t0}ms")
-    } yield {
-      val t2 = System.currentTimeMillis
-      //we want to have nodeStatus for all asked node, not only the ones with reports
-      val nodeStatusReports = runInfos.par.map { case (nodeId, runInfo) =>
-        val status = ExecutionBatch.getNodeStatusReports(nodeId, runInfo, reports.getOrElse(nodeId, Seq()), unexpectedInterpretation)
-        (status.nodeId, status)
+      val agentRunIds = (runBatch.collect { case (nodeId, run: LastRunAvailable) =>
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala]: I'm not sure why we do 2 traverse of runBatch here (I know it was like that before). It seems that we could just do:
```
runBatch.flatMap { case (nodeId, run) => run match {
  case r: LastRunAvailable    => Some(AgentRunId(nodeId, run.lastRunDateTime))
  case Pending(_, Some(r), _) => Some(AgentRunId(nodeId, r._1))
  case _                      => None
} }
```

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala (hunk 1 above)
Comment (ncharles) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala]: yes !

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala (hunk 2 above)
Comment (ncharles) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala]: indeed

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Pull request merged**